### PR TITLE
return_type in {DENSE_GEN, LOD_GEN} allows returning a generator instead of a list

### LINKED ISSUE #83: Support returning a generator (at least for DENSE and LOD)
To allow liac-arff to be memory efficient, but not commit to a particular in-memory representation or libraries, it could have an option which returned the data as an iterator/generator (using `return_type`, or as well as it?). It could either generate one row at a time, or use a [pandas-style `chunksize`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) to deliver multiple rows at a time. Either way it could be used, together with the returned metadata about attributes, to store data in a compact or out-of-memory format on the fly (e.g. using [`numpy.fromiter`](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.fromiter.html)).

### PR DESCRIPTION
**Pull request by jnothman**
Fixes #83 

TODO:
* [x] add tests
* [x] add documentation
* [x] illustrate construction of a pandas DataFrame or a numpy struct array

### CODE REVIEW

**Comment by jnothman** [comment]
Btw, an alternative interface would not create new Data subclasses, and would instead have a parameter return_generator in `Decoder`...

**Comment by mfeurer** [comment]
Sorry, I'm not familiar with your workflow - can I already have a look at this?

**Comment by jnothman** [comment]
Sure: it works. It just isn't tested/documented. (See TODO list above.)

```py
In [6]: import arff
In [8]: out = arff.loads('''@relation foobar
   ...: @attribute foo real
   ...: @attribute bar real
   ...: @data
   ...: 0, 1
   ...: 2, 3
   ...: ''', return_type=arff.DENSE_GEN)

In [9]: out['data']
Out[9]: <generator object DenseGeneratorData.decode_rows at 0x1111b6938>
In [10]: list(out['data'])
Out[10]: [[0.0, 1.0], [2.0, 3.0]]
In [11]: list(out['data'])  # generator is now exhausted
Out[11]: []
```

**Comment by jnothman** [comment]
So the idea of opening a work-in-progress PR is to receive feedback on API design and code structure, while the work may progress.

File: tests/test_data.py [hunk 1]
```diff
@@ -39,8 +39,7 @@ def test_conversor(self):
                       ConversorStub(str if arff.PY3 else unicode)]
 
         fixture = u'Iris,3.4,2,Setosa'
-        self.data.decode_data(fixture, conversors)
-        result = self.data.data[0]
+        result, = self.data.decode_rows([fixture], conversors)
```
Comment (mfeurer) [inline tests/test_data.py]: Just to double-check: the `result,` statement on the left makes sure the generator is exhausted by trying to create a tuple?

Comment (jnothman) [reply tests/test_data.py]: I don't think it creates a tuple as such, but yes, the unpacking syntax is effectively an assertion about the number of elements on the right hand side

**Comment by jnothman** [comment]
I thought the dense reader already handled a mixture of sparse and dense input?

**Comment by mfeurer** [comment]
Indeed, it does. And also WEKA parses that file, very interesting...

Is there anything else I need to do before you commence on the TODOs?

**Comment by jnothman** [comment]
> Is there anything else I need to do before you commence on the TODOs?

No, I just need to find time (or another contributor)

**Comment by mfeurer** [comment]
Looks good to me, but somehow a conflict sneaked in here.

**Pull request merged**